### DISCUSSION TITLE
Sudden high search latencies with 50% used QueryNodes

### DISCUSSION THREAD

**Discussion by Archalbc** [milvus-io/milvus#33293] [Q&A and General discussion] [answered]
Hi !
We are currently facing a quite strange issue on our query nodes (?).
After a lot of investigation and changes we think we have reached a point where no more documentation or Google search could help us to understand what is happenning in our cluster.
Setup:

* Milvus cluster 2.3.12
* 3 proxy nodes, 10 query nodes, 3 index nodes, 2 datanodes
* ~60M entities in 4-5 collections
* query node: 16 CPU, 32GB ram

We got high latencies on our querynodes while their CPU usage was not that high, that's why we tested differents stuff:
* during high traffic, we observed the parallel readTask reaching its maximum value (16 for 16 CPU), so we increased the queryNode.scheduler.maxReadConcurrentRatio to 2
* As we have more than 16GB RAM for querynodes, we increased the max segment size from 512Mo to 2GB.
* We tested adding 1 in-memory replicas for the biggest collection (20M entities)

Timeline of the day:
* 22 May 2024 10:50
  * increase maxConcurrentReadRatio: 2
* 22 May 2024 15:04
  * Add proxies to check if rateLimit or latencies are better
  * Nothing change, back to 3 proxies, no issue with proxyNodes
* 22 May 2024 15:45
  * datacoord.segment.maxSize from 512 to 2048
  * add indexnode count from 2 to 10 to accelerate the re-indexation
* 22 May 2024 16:50-55
  * We are reaching 300 RPS, latencies suddenly explode like at lunch around 13:00

But we don't think we improved the situation that much.

Something quite strange here is that latencies explode as soon as we are approaching 300 RPS. We expect latencies gradually increase with the increase of the traffic load but this is clearly not the case, this is too sudden. (example at lunch 13:00 and at the end of the afternoon at 17:00)

As querynode CPU usage is not high we cannot really understand where is the issue (rate limit? segcore ? queue ? timetick ?), everything suddenly increasing when we reach around 300 RPS.

Bonus: Why we are having rateLimit on DQL whereas this feature is disabled (false) ??

![image](https://github.com/milvus-io/milvus/assets/144429952/413058f6-e6a2-4c10-82e6-5c7622d0ca47)
![image](https://github.com/milvus-io/milvus/assets/144429952/bea026f5-de5d-485b-8860-8ac01bd50aa2)
![image](https://github.com/milvus-io/milvus/assets/144429952/c5ecaa7a-8160-4485-9e72-bd1068efcdf5)
![image](https://github.com/milvus-io/milvus/assets/144429952/f5349392-37bd-4958-90ef-1b522d864cb6)
![image](https://github.com/milvus-io/milvus/assets/144429952/2851f570-0789-45d7-a51f-878f6a34d361)
![image](https://github.com/milvus-io/milvus/assets/144429952/f6a704e1-8717-480f-b73b-b02b18a08354)
![image](https://github.com/milvus-io/milvus/assets/144429952/4a336ae2-2766-4c95-bd9f-8b20762bd5ee)
![image](https://github.com/milvus-io/milvus/assets/144429952/9062f9df-3bd2-4c21-96ee-34c87849526d)
![image](https://github.com/milvus-io/milvus/assets/144429952/2c0e8629-3cf0-4cd1-a2e5-bae82a7ca8c0)

Update:
Why it seems only one querynode got ready readTask ? My collections' segments are well spread.
![image](https://github.com/milvus-io/milvus/assets/144429952/5d7c6cb7-7b87-4fa3-b9ee-c79b399289cf)
legend: {node_id}-{collection_id}  (448835623831330678 is the biggest collection)

![image](https://github.com/milvus-io/milvus/assets/144429952/df80f3c4-b5c7-45f9-a9e3-0b4fc6968938)
![image](https://github.com/milvus-io/milvus/assets/144429952/3e42b5da-e333-4b2b-aec9-451cc53b1a6e)

Thank you for your help.

**Comment by yhmo** [comment]
Some questions:
- what the vector dimension is?
- which type of index?
- the client calls both query() and search() interfaces. what fields are retrieved by query()?

CPU usage of querynode is not high, which indicates the bottleneck is in query() since query() might fetch data from storage.

**Reply by flsworld** [reply]
Hello @yhmo, 

- what the vector dimension is?
256

- which type of index?
HNSW : ef: 100, efContruction: 360, M: 8

- the client calls both query() and search() interfaces. what fields are retrieved by query()?
2 fields : embedding and createdAt

**Reply by yhmo** [reply]
60M 256dim, the raw data size is 60GB. The biggest collection has 20M entities, 20GB.
If segment.maxSize=512MB, ideally this collection has 40 ~ 200 segments. The screenshot shows the number of segments is from 20 to 240 for each query node, far beyond expected.
If segment.maxSize=2GB, ideally this collection has 10 ~ 50 segments. The screenshot shows the number of segments is from 10 to 90 for each query node, far beyond expected.

Does this collection have a **partition key**?

**Reply by Archalbc** [reply]
Hi,

No collection use a partition key.

Here is our current segment allocation
![image](https://github.com/milvus-io/milvus/assets/144429952/33b404c9-694a-4566-a54d-65b1d967056d)

**Comment by xiaofan-luan** [comment]
1. how many shards do you have in the collection? is it always the shard delegator slower than other nodes that slower the overall search? Because from you use cases, it seems that there is usually 1-2 node slower than others.
2. you can try to ignore growing segment see if you can improve the search performance. if yes increase queryCoord.growingRowCountWeight to 20 see if it helps on performance

**Reply by flsworld** [reply]
@Archalbc will be more precise than me. 

1. We got 4 shards per collection
2. Unfortunately this is one of the feature we wanted when opting for milvus : indexing in streaming. We need to search in growing segment..

**Comment by flsworld** [comment]
Hello guys, 

We made some further investigation yesterday. Here is what we discovered : 

During peak latency from 19:30 to 23:00
![Capture d’écran 2024-05-23 à 09 14 04](https://github.com/milvus-io/milvus/assets/99285060/6430317b-1568-423b-93f2-6fb4182b2e9c)

The same querynode **nhbbd** got both its unsolved read requests and read requests queues kinda congested
![Capture d’écran 2024-05-23 à 09 11 02](https://github.com/milvus-io/milvus/assets/99285060/50334053-e685-4225-9045-4c6360d8c800)
![Capture d’écran 2024-05-23 à 09 11 42](https://github.com/milvus-io/milvus/assets/99285060/d8f09b81-a6ff-4d3f-a10a-fff0c980b622)

And its read tasks capped
![Capture d’écran 2024-05-23 à 09 12 38](https://github.com/milvus-io/milvus/assets/99285060/52049d1e-5fb6-4ccb-88bc-31fe04dcdeeb)

So we wanted to see what **nhbbd** had to say and this is what it told us : 
```
=> failed to search: out of range in json: ef(100) should be larger than k(110)
E20240522 17:25:33.143208   498 hnsw_config.h:67] [KNOWHERE][CheckAndAdjustForSearch][milvus] ef(100) should be larger than k(110)
E20240522 17:25:33.151029   453 hnsw_config.h:67] [KNOWHERE][CheckAndAdjustForSearch][milvus] ef(100) should be larger than k(110)
```

I got questions starting from there : 
- is it possible that this single querynode could plumb the overall querynode component 
- when the milvus client asks for ANN, topK is either 20 or 30, never 110.  Is there any mechanisms in the milvus cluster that could increase this value to 110 ?

**Reply by xiaofan-luan** [reply]
From my current knowledge, unless you are using range search, milvus won't automaticlly increase topk.

What is your target QPS?

**Reply by flsworld** [reply]
We may plan to use the range search but it's not the case actually. Our target RPS is 1410 in 6 months. But as for now we'd like to be able to run it with 470 RPS.

**Reply by flsworld** [reply]
Even if we don't know what caused the topK to be increased to 110, we set up the ef to 111 in order to avoid the errors. But the problem remain exactly the same. 


<img width="1406" alt="Capture d’écran 2024-05-23 à 17 04 14" src="https://github.com/milvus-io/milvus/assets/99285060/5296c368-c487-4063-aba1-7d30ae51044a">


<img width="1412" alt="Capture d’écran 2024-05-23 à 16 58 32" src="https://github.com/milvus-io/milvus/assets/99285060/ae0b1928-5098-472f-ac09-7707e58bcc7f">
<img width="1412" alt="Capture d’écran 2024-05-23 à 16 59 17" src="https://github.com/milvus-io/milvus/assets/99285060/e1bfc45a-ea21-4461-8518-57dd5f830f61">
<img width="1411" alt="Capture d’écran 2024-05-23 à 17 01 10" src="https://github.com/milvus-io/milvus/assets/99285060/d98c3506-6076-424d-93d2-74ff18a3ab76">

<img width="1414" alt="Capture d’écran 2024-05-23 à 17 13 21" src="https://github.com/milvus-io/milvus/assets/99285060/4835350f-4ab2-4154-9786-c21c535a3ca3">

<img width="1417" alt="Capture d’écran 2024-05-23 à 17 14 34" src="https://github.com/milvus-io/milvus/assets/99285060/b1669157-4246-4dbd-8cf8-9ecb59482bf2">
<img width="1415" alt="Capture d’écran 2024-05-23 à 17 14 59" src="https://github.com/milvus-io/milvus/assets/99285060/5ef9e3ba-273a-4688-8b82-1f35b9d12ba3">

**Reply by flsworld** [reply]
This is the stack trace of one of the log in the querynode in error during the bump of the latency. The other log messages are `retry` and are failing the same way

```
github.com/milvus-io/milvus/pkg/util/retry.Do
	/go/src/github.com/milvus-io/milvus/pkg/util/retry/retry.go:46
github.com/milvus-io/milvus/internal/util/grpcclient.(*ClientBase[...]).call
	/go/src/github.com/milvus-io/milvus/internal/util/grpcclient/client.go:467
github.com/milvus-io/milvus/internal/util/grpcclient.(*ClientBase[...]).Call
	/go/src/github.com/milvus-io/milvus/internal/util/grpcclient/client.go:553
github.com/milvus-io/milvus/internal/util/grpcclient.(*ClientBase[...]).ReCall
	/go/src/github.com/milvus-io/milvus/internal/util/grpcclient/client.go:569
github.com/milvus-io/milvus/internal/distributed/querynode/client.wrapGrpcCall[...]
	/go/src/github.com/milvus-io/milvus/internal/distributed/querynode/client/client.go:88
github.com/milvus-io/milvus/internal/distributed/querynode/client.(*Client).SearchSegments
	/go/src/github.com/milvus-io/milvus/internal/distributed/querynode/client/client.go:206
github.com/milvus-io/milvus/internal/querynodev2/cluster.(*remoteWorker).SearchSegments
	/go/src/github.com/milvus-io/milvus/internal/querynodev2/cluster/worker.go:107
github.com/milvus-io/milvus/internal/querynodev2/delegator.(*shardDelegator).Search.func3
	/go/src/github.com/milvus-io/milvus/internal/querynodev2/delegator/delegator.go:251
github.com/milvus-io/milvus/internal/querynodev2/delegator.executeSubTasks[...].func1
	/go/src/github.com/milvus-io/milvus/internal/querynodev2/delegator/delegator.go:520
```

**Reply by xiaofan-luan** [reply]
I think this could fail due to timeout. 
your cpu usage is already very high.
need to analyze why cpu is high.
a pprof could be very helpful

**Reply by xiaofan-luan** [reply]
<img width="708" alt="image" src="https://github.com/milvus-io/milvus/assets/83447078/28a8fe52-0f83-42c0-81f3-16ee08b3dbd4">

your cpu request is only 8 core and it seems that many of the cpu already used more than 8 core. 

by saying 16core, you should use set request to 16 core

**Reply by flsworld** [reply]
Yeah thanks a lot @xiaofan-luan ! One of our next move will be to leverage on pprof. You're right about the cpu, we'll increase its value this morning

**Comment by xiaofan-luan** [comment]
obviously, the reason why node is slow is long queue latency
![image](https://github.com/milvus-io/milvus/assets/83447078/26067b3f-a794-4c07-bcf1-1feb74b769d9)

and cpu usage is already high ->  For 16 cpu node, you've used more than 12 cores from your monitoring system.

Next step:
1. do a compaction see how it helps on improving the performance(it will take some time to compact and rebuild index)
2. pprof the busiest node and see what part of the request eats your cpu.
3. scale up or out
4. try zlliz cloud(should be 3-5 times faster under same resources)

**Comment by flsworld** [comment]
I was thinking about the topK value that is not the one we requested. Would it be possible that the query nodes ask for a higher topK depending on the targeted bucket ? 
The average values in `search group topK` and `search topK` are the one requested in the milvus client. The fluctuations would represent some compensations due to the lack of results in a bucket 🤷 

![Capture d’écran 2024-05-24 à 07 30 53](https://github.com/milvus-io/milvus/assets/99285060/f6abea19-975e-4e8d-a463-642456895b05)

![Capture d’écran 2024-05-24 à 07 31 09](https://github.com/milvus-io/milvus/assets/99285060/902ad307-2402-475e-a3f4-340183f15b31)


Except than that, I noticed that in the search topK, there are 2 query nodes that are not contributing : `4895` & `4896`. I was wondering why because it is not the case in the search group topK.

**Reply by xiaofan-luan** [reply]
> I was thinking about the topK value that is not the one we requested. Would it be possible that the query nodes ask for a higher topK depending on the targeted bucket ? The average values in `search group topK` and `search topK` are the one requested in the milvus client. The fluctuations would represent some compensations due to the lack of results in a bucket 🤷
> 
> ![Capture d’écran 2024-05-24 à 07 30 53](https://private-user-images.githubusercontent.com/99285060/333463081-f6abea19-975e-4e8d-a463-642456895b05.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY1MzI3NzksIm5iZiI6MTcxNjUzMjQ3OSwicGF0aCI6Ii85OTI4NTA2MC8zMzM0NjMwODEtZjZhYmVhMTktOTc1ZS00ZThkLWE0NjMtNjQyNDU2ODk1YjA1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTI0VDA2MzQzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTczM2ZhYzMwOTJlYjg1ZDRiMWVlNjRiNzg2OTE2ODdkOGIxZDMwMGIwZWYxZjQ5MWUzNWU0NGQ5ODhlYzUwYjcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.DwNBRXj596c8daYhO6aOM7D8oQc56sKqjjlYiafCZnU)
> 
> ![Capture d’écran 2024-05-24 à 07 31 09](https://private-user-images.githubusercontent.com/99285060/333463119-902ad307-2402-475e-a3f4-340183f15b31.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY1MzI3NzksIm5iZiI6MTcxNjUzMjQ3OSwicGF0aCI6Ii85OTI4NTA2MC8zMzM0NjMxMTktOTAyYWQzMDctMjQwMi00NzVlLWEzZjQtMzQwMTgzZjE1YjMxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTI0VDA2MzQzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA3M2Y4ZTZhMTVhN2Q2ZDA0ZGU4NGU4NjQyNGNiZjEwOGNkY2YwODk3ZjgxZjFlOTEzNDljMWUzNzg0Y2EyM2ImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.s92KUIkVxMvtcZXMFhJ4fzy9Z6ugIViq6APbhb9faMY)
> 
> Except than that, I noticed that in the search topK, there are 2 query nodes that are not contributing : `4895` & `4896`. I was wondering why because it is not the case in the search group topK.

no, milvus won't increase topk .

**Comment by Archalbc** [comment]
Hi!
Here is some news:
* we set cpu request/limit to 16 for querynode
* we set 2 replicas for all collections

Now the load is more spread accross all query nodes and I'm starting to wonder if the bottleneck is the fact that as all collections have 4 shards, there are too many shard delegators and growing segment per collection + 10 query nodes, maybe it's a bit too much of messages to process for querynode delegators

![image](https://github.com/milvus-io/milvus/assets/144429952/ecdffd1b-1e5b-4a2a-bd5d-b4827016adef)
![image](https://github.com/milvus-io/milvus/assets/144429952/06841441-56bc-4c8b-b243-4c5b28058ea7)

Should we recreate all collection with  shards_num=2.
what do you think about that  ?

pprof in something we will do after this move.

**Comment by xiaofan-luan** [comment]
I don't think it's really help to reduce shard numbers, unless querynodes is the bottleneck.

But you could try with next release. 
with 10 nodes, and 2 shards, use ChannelLevelScoreBalancerName will ensure each shard is only spread to 4 nodes.

The segment  number distribution seems to be not very balanced. did you check the the reason?

**Reply by Archalbc** [reply]
Hello,
Yes we checked the distribution and it seems it was a mess because karpenter moved a lot of querynode for placement optimization ><

**Comment by Archalbc** [comment, marked as answer]
Hey !
It seems we finally stabilize this situation and understand a lot of stuff. Here is the summary of everything we did to conclude this issue:

* Add 2 replicas to each collection: Better spread of search on all query nodes
* Fix failed requests from querynode logs
  * "ef(100) should be larger than k(110)"
  * It was an issue on the client app side.
* Set queryNode.scheduler.maxReadConcurrentRatio from 1 to 2: better use of the querynode's CPU
* Increase dataCoord.segment.maxsize from 512 Mo to 2Gb as we have queryNode with 32GB Ram : Divide by 4 the number of segments > Improve the search latency
* Define queryNode pod AntiAffinity to enforce their placement on different Kubernetes worker nodes.
* Increase resource "request" CPU for Query Nodes to have a better guarantee of usable resources. (Resources "limits" stay the same)

Thank a lot again for your help and availability that we really appreciate <3
